### PR TITLE
ESP32 module support

### PR DESCRIPTION
**Pull request by yamt**

### CODE REVIEW

**Comment by yamt** [comment]
@patacongo this pr includes a commit https://github.com/apache/incubator-nuttx/pull/447/commits/243427296613dcbe650138fb248d8009012832f6 which modifies your copyright notice. please review. thank you.

**Comment by masayuki2009** [comment]
Hi, @yamt,

Regarding relocation, could you tell us where we can find the Tensilica ISA document?

File: libs/libc/machine/xtensa/arch_elf.c [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+ *
+ ****************************************************************************/
+
+bool up_checkarch(FAR const Elf32_Ehdr *ehdr)
+{
+  /* Make sure it's an Xtensa executable */
+
+  if (ehdr->e_machine != EM_XTENSA)
+    {
+      berr("ERROR: Not for ARM: e_machine=%04x\n", ehdr->e_machine);
```
Comment (masayuki2009) [inline libs/libc/machine/xtensa/arch_elf.c]: "Not for Xtensa" is correct.

File: libs/libc/machine/xtensa/arch_elf.c [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+ *   value indicating the cause of the relocation failure.
+ *
+ ****************************************************************************/
+
+int up_relocate(FAR const Elf32_Rel *rel, FAR const Elf32_Sym *sym,
+                uintptr_t addr)
+{
+  unsigned int relotype;
+
+  /* All relocations except R_ARM_V4BX depend upon having valid symbol
```
Comment (masayuki2009) [inline libs/libc/machine/xtensa/arch_elf.c]: "except R_ARM_V4BX" should be removed or changed.

Comment (yamt) [reply libs/libc/machine/xtensa/arch_elf.c]: @masayuki2009 
i'm looking at https://github.com/eerimoq/hardware-reference/blob/master/esp32/xtensa%20Instruction%20Set%20Architecture%20(ISA)%20Reference%20Manual.pdf
for instruction encoding.
wrt relocation, i have no documentation. this code is based on some reading of espressif's llvm fork + wild guesses.
https://github.com/espressif/llvm-project

File: libs/libc/machine/xtensa/arch_elf.c (hunk 1 above)
Comment (yamt) [reply libs/libc/machine/xtensa/arch_elf.c]: i will fix when i push next time

File: libs/libc/machine/xtensa/arch_elf.c (hunk 2 above)
Comment (yamt) [reply libs/libc/machine/xtensa/arch_elf.c]: i will fix when i push next time

**Comment by yamt** [comment]
@masayuki2009 
i'm looking at https://github.com/eerimoq/hardware-reference/blob/master/esp32/xtensa%20Instruction%20Set%20Architecture%20(ISA)%20Reference%20Manual.pdf
for instruction encoding.
wrt relocation, i have no documentation. this code is based on some reading of espressif's llvm fork + wild guesses.
https://github.com/espressif/llvm-project

**Comment by masayuki2009** [comment]
@yamt Thanks for the information.

File: include/elf.h [hunk 3]
```diff
@@ -71,6 +71,7 @@
 #define EM_CRIS            76     /* Axis Communications 32-bit embedded processor */
 #define EM_V850            87     /* NEC v850 */
 #define EM_M32R            88     /* Renesas M32R */
+#define EM_XTENSA          94     /* Tensilca Xtensa */
```
Comment (midokura-silvia) [inline include/elf.h]: Should be Tensilica (missing "i")

Comment (yamt) [reply include/elf.h]: i will fix when i push next time

File: libs/libc/machine/xtensa/arch_elf.c [hunk 4]
```diff
@@ -154,6 +161,57 @@ int up_relocate(FAR const Elf32_Rel *rel, FAR const Elf32_Sym *sym,
...
+
+  relotype = ELF32_R_TYPE(rel->r_info);
+  value = sym->st_value + rel->r_addend;
+
+  /* Handle the relocation by relocation type */
+
+  switch (relotype)
+    {
+    case R_XTENSA_32:
+      (*(uint32_t *)addr) += value;
```
Comment (xiaoxiang781216) [inline libs/libc/machine/xtensa/arch_elf.c]: FAR uint32_t *?

File: libs/libc/machine/xtensa/arch_elf.c [hunk 5]
```diff
@@ -154,6 +161,57 @@ int up_relocate(FAR const Elf32_Rel *rel, FAR const Elf32_Sym *sym,
...
+      break;
+
+    case R_XTENSA_ASM_EXPAND:
+      bwarn("ERROR: Ignoring RELA relocation R_XTENSA_ASM_EXPAND %u\n",
+          relotype);
+      break;
+
+    case R_XTENSA_SLOT0_OP:
+      p = (unsigned char *)addr;
```
Comment (xiaoxiang781216) [inline libs/libc/machine/xtensa/arch_elf.c]: FAR unsigned char *

File: libs/libc/machine/xtensa/arch_elf.c (hunk 1 above)
Comment (yamt) [reply libs/libc/machine/xtensa/arch_elf.c]: done

File: libs/libc/machine/xtensa/arch_elf.c (hunk 2 above)
Comment (yamt) [reply libs/libc/machine/xtensa/arch_elf.c]: done

File: include/elf.h (hunk 3 above)
Comment (yamt) [reply include/elf.h]: done

File: libs/libc/machine/xtensa/arch_elf.c (hunk 4 above)
Comment (yamt) [reply libs/libc/machine/xtensa/arch_elf.c]: i will fix when i push next time

File: libs/libc/machine/xtensa/arch_elf.c (hunk 5 above)
Comment (yamt) [reply libs/libc/machine/xtensa/arch_elf.c]: i will fix when i push next time

Comment (yamt) [reply libs/libc/machine/xtensa/arch_elf.c]: done

File: libs/libc/machine/xtensa/arch_elf.c (hunk 4 above)
Comment (yamt) [reply libs/libc/machine/xtensa/arch_elf.c]: done

**Comment by yamt** [comment]
this now works on a real hardware. tested on ESP-EYE.

**Comment by jerpelea** [comment]
+1

**Comment by patacongo** [comment]
Is this PR ready to be merged (when it finishes the current checks).  If there are no open issue then we should take cared of this.  It has been open for some time.  Is there any reason why I should not merge it?

**Comment by yamt** [comment]
@patacongo this part https://github.com/apache/incubator-nuttx/pull/447/commits/6699f5d70f4991a84616a5ce689254b4846e14ec needs your explicit approval.
otherwise there are no known blocking issues.

**Comment by patacongo** [comment]
> @patacongo this part [6699f5d](https://github.com/apache/incubator-nuttx/commit/6699f5d70f4991a84616a5ce689254b4846e14ec) needs your explicit approval.
> otherwise there are no known blocking issues.

Everyone has my explicit approval to change the old BSD heaers to the new Apache headers provided that I am the only person claiming a Copyright.

**Comment by patacongo** [comment]
There are 14 commits for one functional change.  I would say that this PR is a candiate for squashing.

**Comment by yamt** [comment]
@patacongo i can do some squash. but i want to keep some of them separate. eg. https://github.com/apache/incubator-nuttx/pull/447/commits/cd5ea8d891843dc41342590f5a05fdd8b0269922
after all i'm not sure what's wrong with 14.

**Comment by patacongo** [comment]
Okay... let's now worry about squashing.  I will rebase.  i saw only a few commits that would be better merged.

**Pull request merged**